### ISSUE TITLE
Implement PPO training loop

### ISSUE THREAD

**Issue by tgrunnagle** [tgrunnagle/go-towerfall#40]
## Summary

Implement the core PPO (Proximal Policy Optimization) training loop that handles rollout collection, advantage estimation using Generalized Advantage Estimation (GAE), and policy/value network updates with clipped surrogate objective.

**Task ID:** TASK-022
**Parent Task:** TASK-P006 (PPO Agent Implementation)
**Epic:** #14

## Dependencies

- **#39 (TASK-021)**: Implement PPO actor-critic network - Provides `ActorCriticNetwork` class with `get_action_and_value()` and `get_value()` methods

## Background

From the intake requirements:
- PPO is the chosen RL algorithm for training ML bots against the go-towerfall game
- The training pipeline should use separate components following good engineering practices
- The observation space is 114 values normalized to [-1, 1]
- The action space has 27 discrete actions
- The "better" threshold for successive training is based on average kills/deaths ratio

PPO is an on-policy actor-critic algorithm that:
1. Collects rollouts using the current policy
2. Computes advantages using GAE (Generalized Advantage Estimation)
3. Updates policy using a clipped surrogate objective to prevent large policy changes
4. Updates value function to better estimate state values

### 1. Architecture Overview

The PPO trainer consists of:
1. **RolloutBuffer**: Stores experience during rollout collection
2. **PPOTrainer**: Main training class that orchestrates collection and updates
3. **GAE computation**: Calculates advantages from rewards and values

### 2. File Structure

```
bot2/
  agent/
    __init__.py
    network.py          # From TASK-021
    ppo_trainer.py      # PPO training loop (this task)
    rollout_buffer.py   # Rollout storage
  tests/
    test_ppo_trainer.py # Unit tests
```

#### 3.1 RolloutBuffer Class

Stores experience tuples during rollout collection:

```python
import torch
import numpy as np
from dataclasses import dataclass

@dataclass
class RolloutBuffer:
    """Buffer for storing rollout experience."""

    observations: torch.Tensor      # (steps, obs_size)
    actions: torch.Tensor           # (steps,)
    log_probs: torch.Tensor         # (steps,)
    rewards: torch.Tensor           # (steps,)
    values: torch.Tensor            # (steps,)
    dones: torch.Tensor             # (steps,)
    advantages: torch.Tensor | None = None  # Computed after rollout
    returns: torch.Tensor | None = None     # Computed after rollout

    @classmethod
    def create(cls, num_steps: int, observation_size: int, device: torch.device) -> "RolloutBuffer":
        """Create an empty buffer with pre-allocated tensors."""
        return cls(
            observations=torch.zeros(num_steps, observation_size, device=device),
            actions=torch.zeros(num_steps, dtype=torch.long, device=device),
            log_probs=torch.zeros(num_steps, device=device),
            rewards=torch.zeros(num_steps, device=device),
            values=torch.zeros(num_steps, device=device),
            dones=torch.zeros(num_steps, device=device),
        )

    def compute_advantages(
        self,
        last_value: torch.Tensor,
        last_done: torch.Tensor,
        gamma: float = 0.99,
        gae_lambda: float = 0.95,
    ) -> None:
        """Compute GAE advantages and returns."""
        num_steps = self.rewards.shape[0]
        advantages = torch.zeros_like(self.rewards)
        last_gae = 0.0

        for t in reversed(range(num_steps)):
            if t == num_steps - 1:
                next_non_terminal = 1.0 - last_done.float()
                next_value = last_value
            else:
                next_non_terminal = 1.0 - self.dones[t + 1]
                next_value = self.values[t + 1]

            delta = self.rewards[t] + gamma * next_value * next_non_terminal - self.values[t]
            advantages[t] = last_gae = delta + gamma * gae_lambda * next_non_terminal * last_gae

        self.advantages = advantages
        self.returns = advantages + self.values

    def get_minibatches(
        self,
        batch_size: int,
        shuffle: bool = True,
    ) -> list[dict[str, torch.Tensor]]:
        """Split buffer into minibatches for training."""
        num_steps = self.rewards.shape[0]
        indices = np.arange(num_steps)
        if shuffle:
            np.random.shuffle(indices)

        batches = []
        for start in range(0, num_steps, batch_size):
            end = min(start + batch_size, num_steps)
            batch_indices = indices[start:end]
            batches.append({
                "observations": self.observations[batch_indices],
                "actions": self.actions[batch_indices],
                "log_probs": self.log_probs[batch_indices],
                "advantages": self.advantages[batch_indices],
                "returns": self.returns[batch_indices],
                "values": self.values[batch_indices],
            })
        return batches
```

#### 3.2 PPOTrainer Class

Main training class:

```python
import torch
import torch.nn as nn
import torch.optim as optim
from dataclasses import dataclass
from typing import Protocol

from bot2.agent.network import ActorCriticNetwork
from bot2.agent.rollout_buffer import RolloutBuffer


class Environment(Protocol):
    """Protocol for gym-like environment."""
    def reset(self) -> tuple[torch.Tensor, dict]: ...
    def step(self, action: int) -> tuple[torch.Tensor, float, bool, bool, dict]: ...


@dataclass
class PPOConfig:
    """PPO hyperparameters."""
    # Rollout
    num_steps: int = 2048           # Steps per rollout

    # GAE
    gamma: float = 0.99             # Discount factor
    gae_lambda: float = 0.95        # GAE lambda

    # PPO
    num_epochs: int = 10            # Epochs per update
    minibatch_size: int = 64        # Minibatch size
    clip_range: float = 0.2         # PPO clip range (epsilon)
    clip_range_vf: float | None = None  # Value function clip range (None = no clipping)

    # Loss coefficients
    value_coef: float = 0.5         # Value loss coefficient
    entropy_coef: float = 0.01      # Entropy bonus coefficient

    # Optimization
    learning_rate: float = 3e-4     # Learning rate
    max_grad_norm: float = 0.5      # Gradient clipping norm

    # Normalization
    normalize_advantages: bool = True


class PPOTrainer:
    """PPO training loop implementation."""

    def __init__(
        self,
        network: ActorCriticNetwork,
        config: PPOConfig | None = None,
        device: torch.device | None = None,
    ):
        self.config = config or PPOConfig()
        self.device = device or torch.device("cuda" if torch.cuda.is_available() else "cpu")
        self.network = network.to(self.device)
        self.optimizer = optim.Adam(
            self.network.parameters(),
            lr=self.config.learning_rate,
        )

        # Training statistics
        self.total_timesteps = 0
        self.num_updates = 0

    def collect_rollout(
        self,
        env: Environment,
        obs: torch.Tensor,
    ) -> tuple[RolloutBuffer, torch.Tensor, bool]:
        """
        Collect rollout experience from environment.

        Args:
            env: Gym-like environment
            obs: Current observation

        Returns:
            buffer: Filled rollout buffer
            next_obs: Observation after rollout
            done: Whether episode ended
        """
        buffer = RolloutBuffer.create(
            self.config.num_steps,
            obs.shape[-1],
            self.device,
        )

        done = False
        for step in range(self.config.num_steps):
            obs_tensor = obs.to(self.device)

            with torch.no_grad():
                action, log_prob, _, value = self.network.get_action_and_value(
                    obs_tensor.unsqueeze(0)
                )

            # Store experience
            buffer.observations[step] = obs_tensor
            buffer.actions[step] = action.squeeze()
            buffer.log_probs[step] = log_prob.squeeze()
            buffer.values[step] = value.squeeze()

            # Take action in environment
            next_obs, reward, terminated, truncated, info = env.step(action.item())
            done = terminated or truncated

            buffer.rewards[step] = reward
            buffer.dones[step] = float(done)

            obs = torch.as_tensor(next_obs, dtype=torch.float32)

            if done:
                obs, _ = env.reset()
                obs = torch.as_tensor(obs, dtype=torch.float32)

        # Compute advantages with bootstrapped value
        with torch.no_grad():
            last_value = self.network.get_value(obs.to(self.device).unsqueeze(0))

        buffer.compute_advantages(
            last_value.squeeze(),
            torch.tensor(done, device=self.device),
            self.config.gamma,
            self.config.gae_lambda,
        )

        self.total_timesteps += self.config.num_steps
        return buffer, obs, done

    def update(self, buffer: RolloutBuffer) -> dict[str, float]:
        """
        Perform PPO update on collected rollout.

        Args:
            buffer: Rollout buffer with computed advantages

        Returns:
            Dictionary of training metrics
        """
        # Track metrics
        policy_losses = []
        value_losses = []
        entropy_losses = []
        clip_fractions = []
        approx_kls = []

        for epoch in range(self.config.num_epochs):
            for batch in buffer.get_minibatches(self.config.minibatch_size):
                # Get current policy outputs
                _, new_log_prob, entropy, new_value = self.network.get_action_and_value(
                    batch["observations"],
                    batch["actions"],
                )

                # Compute advantages (optionally normalize)
                advantages = batch["advantages"]
                if self.config.normalize_advantages and advantages.numel() > 1:
                    advantages = (advantages - advantages.mean()) / (advantages.std() + 1e-8)

                # Policy loss with clipping
                log_ratio = new_log_prob - batch["log_probs"]
                ratio = torch.exp(log_ratio)

                # Approximate KL divergence for monitoring
                with torch.no_grad():
                    approx_kl = ((ratio - 1) - log_ratio).mean()
                    clip_fraction = ((ratio - 1.0).abs() > self.config.clip_range).float().mean()

                policy_loss_1 = -advantages * ratio
                policy_loss_2 = -advantages * torch.clamp(
                    ratio,
                    1.0 - self.config.clip_range,
                    1.0 + self.config.clip_range,
                )
                policy_loss = torch.max(policy_loss_1, policy_loss_2).mean()

                # Value loss (optionally clipped)
                if self.config.clip_range_vf is not None:
                    value_clipped = batch["values"] + torch.clamp(
                        new_value - batch["values"],
                        -self.config.clip_range_vf,
                        self.config.clip_range_vf,
                    )
                    value_loss_1 = (new_value - batch["returns"]) ** 2
                    value_loss_2 = (value_clipped - batch["returns"]) ** 2
                    value_loss = 0.5 * torch.max(value_loss_1, value_loss_2).mean()
                else:
                    value_loss = 0.5 * ((new_value - batch["returns"]) ** 2).mean()

                # Entropy loss (negative because we want to maximize entropy)
                entropy_loss = -entropy.mean()

                # Total loss
                loss = (
                    policy_loss
                    + self.config.value_coef * value_loss
                    + self.config.entropy_coef * entropy_loss
                )

                # Optimization step
                self.optimizer.zero_grad()
                loss.backward()
                nn.utils.clip_grad_norm_(
                    self.network.parameters(),
                    self.config.max_grad_norm,
                )
                self.optimizer.step()

                # Record metrics
                policy_losses.append(policy_loss.item())
                value_losses.append(value_loss.item())
                entropy_losses.append(-entropy_loss.item())
                clip_fractions.append(clip_fraction.item())
                approx_kls.append(approx_kl.item())

        self.num_updates += 1

        return {
            "policy_loss": np.mean(policy_losses),
            "value_loss": np.mean(value_losses),
            "entropy": np.mean(entropy_losses),
            "clip_fraction": np.mean(clip_fractions),
            "approx_kl": np.mean(approx_kls),
            "total_timesteps": self.total_timesteps,
            "num_updates": self.num_updates,
        }

    def train_step(
        self,
        env: Environment,
        obs: torch.Tensor,
    ) -> tuple[dict[str, float], torch.Tensor]:
        """
        Perform one full training step (rollout + update).

        Args:
            env: Gym-like environment
            obs: Current observation

        Returns:
            metrics: Dictionary of training metrics
            next_obs: Observation after rollout
        """
        buffer, next_obs, _ = self.collect_rollout(env, obs)
        metrics = self.update(buffer)
        return metrics, next_obs
```

#### 4.1 Generalized Advantage Estimation (GAE)
GAE balances bias and variance in advantage estimation:
- `gamma`: Discount factor (0.99) - how much to value future rewards
- `gae_lambda`: GAE parameter (0.95) - interpolates between TD(0) and Monte Carlo
- Formula: `A_t = sum_{l=0}^{inf} (gamma * lambda)^l * delta_{t+l}`

#### 4.2 Clipped Surrogate Objective
Prevents large policy updates:
- Ratio: `r_t = pi(a|s) / pi_old(a|s)`
- Clipped objective: `min(r_t * A_t, clip(r_t, 1-eps, 1+eps) * A_t)`
- Clip range `eps` = 0.2 is standard

#### 4.3 Value Function Loss
- MSE between predicted value and computed returns
- Optional clipping similar to policy (often disabled)

#### 4.4 Entropy Bonus
- Encourages exploration by adding entropy to the objective
- Coefficient typically 0.01

### 5. Integration Points

- **ActorCriticNetwork (TASK-021)**: Used for policy/value inference and training
- **PPOConfig (TASK-024)**: Will integrate with hyperparameter configuration system
- **Gym Environment (TASK-014)**: Will work with the gymnasium environment wrapper
- **Training Orchestrator (TASK-027)**: Will call `train_step()` in the main training loop
- **Metrics Logging (TASK-031)**: Will consume the metrics dictionary from `update()`

### 6. Usage Example

```python
from bot2.agent.network import ActorCriticNetwork
from bot2.agent.ppo_trainer import PPOTrainer, PPOConfig

# Create network and trainer
network = ActorCriticNetwork(observation_size=114, action_size=27)
config = PPOConfig(
    num_steps=2048,
    num_epochs=10,
    minibatch_size=64,
    learning_rate=3e-4,
)
trainer = PPOTrainer(network, config)

# Training loop (simplified - real version uses gym env)
obs, _ = env.reset()
obs = torch.as_tensor(obs, dtype=torch.float32)

for iteration in range(num_iterations):
    metrics, obs = trainer.train_step(env, obs)
    print(f"Iteration {iteration}: {metrics}")
```

## Acceptance Criteria

- [ ] `RolloutBuffer` class implemented in `bot2/agent/rollout_buffer.py`
  - [ ] Pre-allocates tensors for efficient storage
  - [ ] `compute_advantages()` implements GAE correctly
  - [ ] `get_minibatches()` returns shuffled minibatches
- [ ] `PPOConfig` dataclass with all hyperparameters
- [ ] `PPOTrainer` class implemented in `bot2/agent/ppo_trainer.py`
  - [ ] `collect_rollout()` gathers experience from environment
  - [ ] `update()` performs PPO policy and value updates
  - [ ] `train_step()` combines rollout and update
- [ ] PPO clipped surrogate objective implemented correctly
- [ ] Value loss computed (with optional clipping)
- [ ] Entropy bonus added to loss
- [ ] Gradient clipping applied
- [ ] Advantage normalization (optional, configurable)
- [ ] Training metrics returned from `update()`
- [ ] Unit tests verify:
  - [ ] GAE computation produces correct values
  - [ ] Clipping is applied when ratio exceeds bounds
  - [ ] Minibatch iteration covers all samples
  - [ ] Metrics are computed correctly
- [ ] Type hints using native Python types per CLAUDE.md guidelines
- [ ] Imports relative to bot2 directory

## Test Cases

```python
def test_gae_computation():
    """Test GAE produces expected advantages."""
    buffer = RolloutBuffer.create(5, 10, torch.device("cpu"))
    buffer.rewards = torch.tensor([1.0, 0.0, 1.0, 0.0, 0.0])
    buffer.values = torch.tensor([0.5, 0.5, 0.5, 0.5, 0.5])
    buffer.dones = torch.tensor([0.0, 0.0, 0.0, 0.0, 0.0])

    buffer.compute_advantages(
        last_value=torch.tensor(0.5),
        last_done=torch.tensor(0.0),
        gamma=0.99,
        gae_lambda=0.95,
    )

    assert buffer.advantages is not None
    assert buffer.returns is not None
    assert buffer.advantages.shape == (5,)
    # Verify advantages decrease as we move back in time from rewards

def test_minibatch_coverage():
    """Test all samples are covered in minibatches."""
    buffer = RolloutBuffer.create(100, 10, torch.device("cpu"))
    buffer.advantages = torch.randn(100)
    buffer.returns = torch.randn(100)

    batches = buffer.get_minibatches(batch_size=32)

    total_samples = sum(b["observations"].shape[0] for b in batches)
    assert total_samples == 100

def test_ppo_update_reduces_loss():
    """Test that PPO update actually trains the network."""
    network = ActorCriticNetwork(observation_size=10, action_size=5)
    trainer = PPOTrainer(network, PPOConfig(num_steps=64, num_epochs=5))

    # Create mock buffer with clear signal
    buffer = RolloutBuffer.create(64, 10, trainer.device)
    buffer.observations = torch.randn(64, 10)
    buffer.actions = torch.randint(0, 5, (64,))
    buffer.log_probs = torch.randn(64)
    buffer.rewards = torch.randn(64)
    buffer.values = torch.randn(64)
    buffer.dones = torch.zeros(64)
    buffer.compute_advantages(torch.tensor(0.0), torch.tensor(0.0))

    metrics = trainer.update(buffer)

    assert "policy_loss" in metrics
    assert "value_loss" in metrics
    assert "entropy" in metrics
    assert metrics["num_updates"] == 1

def test_clip_fraction_computed():
    """Test clip fraction metric is in valid range."""
    network = ActorCriticNetwork()
    trainer = PPOTrainer(network)
    # ... setup buffer and update
    metrics = trainer.update(buffer)
    assert 0.0 <= metrics["clip_fraction"] <= 1.0
```

## Notes

- TASK-024 (hyperparameter configuration) will provide a unified config system that PPOConfig can integrate with
- TASK-027 (training orchestrator) will wrap this trainer in the main training loop
- TASK-031 (metrics logging) will log the metrics dictionary to TensorBoard or files
- Consider adding learning rate annealing schedule in future iteration
- Consider adding early stopping based on KL divergence threshold

## References

- PPO Paper: https://arxiv.org/abs/1707.06347
- PPO Implementation Details: https://iclr-blog-track.github.io/2022/03/25/ppo-implementation-details/
- CleanRL PPO: https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ppo.py
- Stable-Baselines3 PPO: https://stable-baselines3.readthedocs.io/en/master/modules/ppo.html
